### PR TITLE
feat(0021): refine protocol semantics

### PR DESCRIPTION
**Pull request by keynslug**
Things that supposedly should be made clearer to users of this feature eventually.

* Requirements and restrictions of metadata
* Meaning of `PUBACK`s
* Durability
* Interaction with session lifetimes

Also simplify diagrams by removing implementation details that have drifted away from current implementation anyway.

### CODE REVIEW

File: active/0021-transfer-files-over-mqtt.md [hunk 1]
```diff
@@ -6,15 +6,15 @@
 
 ## Abstract
 
-This documents defines protocol to send files from MQTT clients to MQTT server. It is using only PUBLISH and PUBACK messages and does not require MQTT 5.0 features like topic aliases.
+This document defines protocol to send files from MQTT clients to MQTT server. It is using only `PUBLISH` and `PUBACK` messages and does not require MQTT 5.0 features like topic aliases.
```
Comment (savonarola) [inline active/0021-transfer-files-over-mqtt.md:9]: AFAIK RC code in `PUBACK` is v5 feature

File: active/0021-transfer-files-over-mqtt.md [hunk 2]
```diff
@@ -70,42 +70,68 @@ As an example of existing implementation we can look at AWS IoT Core [which prov
...
-* If checksum is included in the `init` message payload, the Broker MUST use it to verify integrity of the file after receving the `fin` message for the corresponding file transfer
+* If checksum is included in the `init` message payload, the Broker MUST use it to verify integrity of the file after receiving the `fin` message for the corresponding file transfer
 * If checksum is included in topic name, Broker MUST use it to verify integrity of corresponding data:
   * segment, if it's a segment transfer message
   * whole file, if it's a `fin` message
 * If checksum verification fails, Broker MUST reject the corresponding data
-* Client MUST use $file Topic to transfer files
-* Broker MUST NOT let clients subscribe to $file topics
+* Client MUST use Topic starting with `$file/` to transfer files
+* Broker MUST NOT let clients subscribe to Topics starting with `$file/` topics
```
Comment (savonarola) [inline active/0021-transfer-files-over-mqtt.md:83]: We should not forget to implement this

Comment (savonarola) [reply active/0021-transfer-files-over-mqtt.md:83]: https://emqx.atlassian.net/browse/EMQX-9772

**Review by savonarola — DISMISSED** [review]

File: active/0021-transfer-files-over-mqtt.md [hunk 3]
```diff
@@ -6,15 +6,15 @@
 
 ## Abstract
 
-This documents defines protocol to send files from MQTT clients to MQTT server. It is using only PUBLISH and PUBACK messages and does not require MQTT 5.0 features like topic aliases.
+This document defines protocol to send files from MQTT clients to MQTT server. It is using only `PUBLISH` and `PUBACK` messages, and requires only the _`PUBACK` Reason Code_ feature which is specific to MQTT 5.0.
```
Comment (id) [inline active/0021-transfer-files-over-mqtt.md:9]: Why `requires`? MQTT < 5 implementations would still work, albeit not as efficient. Or?

File: active/0021-transfer-files-over-mqtt.md [hunk 4]
```diff
@@ -70,42 +70,68 @@ As an example of existing implementation we can look at AWS IoT Core [which prov
...
-  2. $file/{fileId}/{offset}[/{checksum}]
-  3. $file/{fileId}/{offset}[/{checksum}]
-  4. ...
-  3. $file/{fileId}/[fin[/{checksum}] | abort]
+
+### Protocol flow
+
+Data is transferred in PUBLISH packets in the following order:
+
+1. `$file/{fileId}/init`
```
Comment (id) [inline active/0021-transfer-files-over-mqtt.md:90]: Above we say "If checksum is included in the `init` message payload", but here we do not show that.

File: active/0021-transfer-files-over-mqtt.md [hunk 5]
```diff
@@ -70,42 +70,68 @@ As an example of existing implementation we can look at AWS IoT Core [which prov
...
+2. `$file/{fileId}/{offset}[/{checksum}]`
+3. `$file/{fileId}/{offset}[/{checksum}]`
+4. ...
+5. `$file/{fileId}/[fin/{fileSize}[/{checksum}] | abort]`
 
 #### `$file/{fileId}/init` message
 
-Initialize the file transfer. Server is expected to store metadata from the payload in the session along with `{fileId}` as a reference for the rest of file metadata.
+Initialize the file transfer. Server is expected to store metadata from the payload in the session along with `{fileId}` as a reference for the rest of file transfer.
```
Comment (id) [inline active/0021-transfer-files-over-mqtt.md:98]: May be we should make it more clear that `clientId` + `fileId` acts as a primary key for file transfers on the cluster level, and `fileId` is like a primary key for transfers from the same client.

File: active/0021-transfer-files-over-mqtt.md (hunk 4 above)
Comment (id) [reply active/0021-transfer-files-over-mqtt.md:90]: Oh right. But yeah, if we could do this without blowing it up out of proportions - would be nice indeed.

File: active/0021-transfer-files-over-mqtt.md (hunk 3 above)
Comment (keynslug) [reply active/0021-transfer-files-over-mqtt.md]: Right, original wording was more accurate. Backed out last commit.

**Review by id — APPROVED** [review]

**Pull request merged**